IllegalStateException when handling `SyncCommittee` messages

```


2023-09-20 12:03:01 | 2023-09-20 10:03:01.269 WARN  - Encountered exception while processing message for topic /eth2/bba4da96/sync_committee_1/ssz_snappy |  
-- | -- | --
  |   | 2023-09-20 12:03:01 | java.util.concurrent.CompletionException: java.lang.IllegalStateException: Already added 102 |  
  |   | 2023-09-20 12:03:01 | at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?] |  
  |   | 2023-09-20 12:03:01 | at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?] |  
  |   | 2023-09-20 12:03:01 | at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?] |  
  |   | 2023-09-20 12:03:01 | at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] |  
  |   | 2023-09-20 12:03:01 | at java.util.concurrent.CompletableFuture.complete(Unknown Source) ~[?:?] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.statetransition.validation.signatures.AggregatingSignatureVerificationService$SignatureTask.lambda$completeAsync$0(AggregatingSignatureVerificationService.java:259) ~[teku-ethereum-statetransition-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:82) ~[teku-infrastructure-async-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAsync$2(AsyncRunner.java:47) ~[teku-infrastructure-async-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:74) ~[teku-infrastructure-async-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:124) ~[teku-infrastructure-async-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?] |  
  |   | 2023-09-20 12:03:01 | at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?] |  
  |   | 2023-09-20 12:03:01 | at java.lang.Thread.run(Unknown Source) [?:?] |  
  |   | 2023-09-20 12:03:01 | Caused by: java.lang.IllegalStateException: Already added 102 |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessagePool$ContributionData.add(SyncCommitteeMessagePool.java:180) ~[teku-ethereum-statetransition-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessagePool.lambda$doAdd$4(SyncCommitteeMessagePool.java:106) ~[teku-ethereum-statetransition-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at it.unimi.dsi.fastutil.ints.IntSets$Singleton.forEach(IntSets.java:169) ~[fastutil-8.5.12.jar:?] |  
  |   | 2023-09-20 12:03:01 | at it.unimi.dsi.fastutil.ints.IntIterable.forEach(IntIterable.java:146) ~[fastutil-8.5.12.jar:?] |  
  |   | 2023-09-20 12:03:01 | at it.unimi.dsi.fastutil.ints.AbstractIntCollection.forEach(AbstractIntCollection.java:153) ~[fastutil-8.5.12.jar:?] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessagePool.doAdd(SyncCommitteeMessagePool.java:100) ~[teku-ethereum-statetransition-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessagePool.lambda$add$1(SyncCommitteeMessagePool.java:84) ~[teku-ethereum-statetransition-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$thenPeek$37(SafeFuture.java:493) ~[teku-infrastructure-async-23.8.0.jar:23.8.0] |  
  |   | 2023-09-20 12:03:01 | ... 11 more


```

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePool.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePool.java
@@ -25,6 +25,8 @@ import java.util.NavigableMap;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.TreeMap;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes32;
 import tech.pegasys.teku.bls.BLS;
 import tech.pegasys.teku.bls.BLSSignature;
@@ -40,6 +42,9 @@ import tech.pegasys.teku.statetransition.OperationAddedSubscriber;
 import tech.pegasys.teku.statetransition.validation.InternalValidationResult;
 
 public class SyncCommitteeMessagePool implements SlotEventsChannel {
+
+  private static final Logger LOG = LogManager.getLogger();
+
   private final Subscribers<OperationAddedSubscriber<ValidatableSyncCommitteeMessage>> subscribers =
       Subscribers.create(true);
 
@@ -176,10 +181,13 @@ public class SyncCommitteeMessagePool implements SlotEventsChannel {
       IntIterator iterator = participationIndices.iterator();
       while (iterator.hasNext()) {
         int index = iterator.nextInt();
-        if (!this.participationIndices.add(index)) {
-          throw new IllegalStateException("Already added " + index);
+        if (this.participationIndices.add(index)) {
+          this.signatures.add(signature);
+        } else {
+          LOG.debug(
+              "Ignoring already aggregated signature from subcommittee participant index = {}",
+              participationIndices);
         }
-        this.signatures.add(signature);
       }
     }
 

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePool.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePool.java
@@ -185,7 +185,8 @@ public class SyncCommitteeMessagePool implements SlotEventsChannel {
           this.signatures.add(signature);
         } else {
           LOG.trace(
-              "Ignoring already aggregated signature from subcommittee participant index = {}", index);
+              "Ignoring already aggregated signature from subcommittee participant index = {}",
+              index);
         }
       }
     }

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePool.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePool.java
@@ -180,13 +180,12 @@ public class SyncCommitteeMessagePool implements SlotEventsChannel {
     public void add(final IntSet participationIndices, final BLSSignature signature) {
       IntIterator iterator = participationIndices.iterator();
       while (iterator.hasNext()) {
-        int index = iterator.nextInt();
+        final int index = iterator.nextInt();
         if (this.participationIndices.add(index)) {
           this.signatures.add(signature);
         } else {
           LOG.debug(
-              "Ignoring already aggregated signature from subcommittee participant index = {}",
-              participationIndices);
+              "Ignoring already aggregated signature from subcommittee participant index = {}", index);
         }
       }
     }

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePool.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePool.java
@@ -184,7 +184,7 @@ public class SyncCommitteeMessagePool implements SlotEventsChannel {
         if (this.participationIndices.add(index)) {
           this.signatures.add(signature);
         } else {
-          LOG.debug(
+          LOG.trace(
               "Ignoring already aggregated signature from subcommittee participant index = {}", index);
         }
       }

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePoolTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessagePoolTest.java
@@ -247,6 +247,34 @@ class SyncCommitteeMessagePoolTest {
   }
 
   @Test
+  public void
+      shouldNotAggregateSignaturesFromRepeatedMessagesWhenSameSubcommitteeParticipationId() {
+    final SyncCommitteeMessage syncCommitteeMessage =
+        dataStructureUtil.randomSyncCommitteeMessage();
+
+    final ValidatableSyncCommitteeMessage message =
+        ValidatableSyncCommitteeMessage.fromNetwork(syncCommitteeMessage, 1);
+    message.setSubcommitteeAssignments(
+        SyncSubcommitteeAssignments.builder().addAssignment(1, 1).build());
+
+    addValidRemote(message);
+    // Duplicated message, should not aggregate signature a second time
+    addValidRemote(message);
+
+    final BLSSignature signature = message.getMessage().getSignature();
+    // Single signature expected on aggregate
+    final BLSSignature expectedAggregate = BLS.aggregate(List.of(signature));
+
+    final UInt64 slot = message.getSlot();
+    final Bytes32 blockRoot = message.getBeaconBlockRoot();
+
+    final Optional<SyncCommitteeContribution> contribution =
+        pool.createContribution(slot, blockRoot, 1);
+    assertThat(contribution).isPresent();
+    assertThat(contribution.orElseThrow().getSignature()).isEqualTo(expectedAggregate);
+  }
+
+  @Test
   void shouldAggregateSignatureMultipleTimesWhenValidatorInSameSubcommitteeMultipleTimes() {
     final ValidatableSyncCommitteeMessage message =
         ValidatableSyncCommitteeMessage.fromValidator(

